dsgrid/config/project_config.py

```diff
@@ -730,46 +781,204 @@ def _iter_datasets_by_status(self, status: DatasetRegistryStatus) -> InputDatase
-            table_columns.update(table.columns)
-
-        all_dimensions = {d.value for d in DimensionType if d != DimensionType.TIME}
-        missing_dimensions = all_dimensions.difference(table_columns)
-        for dim in missing_dimensions:
-            table_count = 0 if table is None else table.count()
-            other = (
-                self.get_base_dimension(DimensionType(dim))
-                .get_records_dataframe()
-                .selectExpr(f"id as {dim}")
+        if try_load_cache:
+            table = try_load_dimension_associations(self.config_id, dataset_id, pivoted_dimension)
+        else:
+            table = None
+        if table is None:
+            logger.info(
+                "No dimension associations table exists for project_id=%s dataset_id=%s "
+                "pivoted_dimension=%s. Build table and save it. try_load_cache=%s",
+                self.config_id,
+                dataset_id,
+                pivoted_dimension,
+                try_load_cache,
             )
-            if table is None:
-                table = other
+            table = self._make_dimension_association_table(dataset_id, pivoted_dimension)
+            if os.environ.get("__DSGRID_SKIP_SAVING_DIMENSION_ASSOCIATIONS__") is not None:
+                logger.warning("Skip saving dimension associations.")
             else:
-                table = table.crossJoin(other)
-                assert table.count() == other.count() * table_count
-
+                # We expect these tables to be very small in files because of compression.
+                table_name = save_dimension_associations(
+                    table.coalesce(1), self.config_id, dataset_id, pivoted_dimension
+                )
+                # It will be much faster to load the persisted table.
+                table = load_stored_table(table_name)
+        else:
+            logger.info(
+                "Loaded cached dimension associations for %s dataset_id=%s pivoted_dimension=%s",
+                self.config_id,
+                dataset_id,
+                pivoted_dimension,
+            )
         return table
 
+    @track_timing(timer_stats_collector)
+    def _make_dimension_association_table(self, dataset_id, pivoted_dimension):
+        required_dimensions = self.get_dataset(dataset_id).required_dimensions
+        supp_dim_query_names = self.get_supplemental_dimension_to_query_name_mapping()
+        multi_dfs = self._build_multi_dim_requirement_associations(
+            required_dimensions.multi_dimensional,
+            pivoted_dimension,
+            supp_dim_query_names,
+        )
+
+        # Project config construction asserts that there is no intersection of dimensions in
+        # multi and single.
+        existing = set()
+        if pivoted_dimension is not None:
+            existing.add(pivoted_dimension.value)
+        for df in multi_dfs:
+            existing.update(set(df.columns))
```

Comment:
Is `existing` getting used later? Did I miss it?

Reply:
line 851